src/celeritas/io/EventWriter.cc

```diff
@@ -0,0 +1,189 @@
+    CELER_EXPECT(particles_);
+
+    CELER_LOG(info) << "Creating " << to_cstring(fmt) << " event file at "
+                    << filename;
+    ScopedTimeAndRedirect temp_{"HepMC3"};
+
+    if (fmt != Format::hepmc3)
+    {
+        CELER_LOG(warning) << "Writing to a event file format other than "
+                              "hepmc3 may result in an unreadable file";
+    }
+
+    writer_ = [&]() -> std::shared_ptr<HepMC3::Writer> {
+        switch (fmt)
+        {
+            case Format::hepevt:
+                return std::make_shared<HepMC3::WriterHEPEVT>(filename);
+            case Format::hepmc2:
+                return std::make_shared<HepMC3::WriterAsciiHepMC2>(filename);
+            case Format::hepmc3:
+                return std::make_shared<HepMC3::WriterAscii>(filename);
+            default:
+                CELER_ASSERT_UNREACHABLE();
+        }
+    }();
+
+    CELER_ENSURE(writer_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write all primaries from a single event.
+ */
+void EventWriter::operator()(argument_type primaries)
+{
+    CELER_EXPECT(!primaries.empty());
+
+    std::set<EventId::size_type> mismatched_events;
+
+    HepMC3::GenEvent evt;
+    evt.set_units(HepMC3::Units::MEV, HepMC3::Units::CM);
+
+    // Vertex and corresponding celeritas position
+    HepMC3::GenVertexPtr vtx;
+    Primary const* vtx_primary{nullptr};
+
+    // Loop over all primaries
+    for (Primary const& p : primaries)
+    {
+        auto par = std::make_shared<HepMC3::GenParticle>();
+
+        par->set_pid(particles_->id_to_pdg(p.particle_id).get());
+        par->set_status(4);  // "incoming beam particle" from HepMC2 manual
+
+        if (!vtx_primary || vtx_primary->time != p.time
+            || vtx_primary->position != p.position)
+        {
+            if (vtx)
+            {
+                evt.add_vertex(vtx);
+            }
+            // Position has changed: add a new vertex
+            vtx_primary = &p;
+
+            HepMC3::FourVector pos;
+            pos.set_x(p.position[0]);
+            pos.set_y(p.position[1]);
+            pos.set_z(p.position[2]);
+            pos.set_t(p.time / units::centimeter * constants::c_light);
+            vtx = std::make_shared<HepMC3::GenVertex>(pos);
+        }
+
+        HepMC3::FourVector mom;
+        mom.set_px(p.direction[0]);
+        mom.set_py(p.direction[1]);
+        mom.set_pz(p.direction[2]);
+        mom.set_e(value_as<units::MevEnergy>(p.energy));
+        par->set_momentum(mom);
+
+        if (CELER_UNLIKELY(p.event_id != EventId{event_count_}))
+        {
+            mismatched_events.insert(p.event_id.unchecked_get());
+        }
+
+        // Note: primary's track ID is ignored
+
+        // Note: the particle has to be added to both in *and* out to be
+        // readable in hepmc3 format
+        CELER_ASSERT(vtx);
+        vtx->add_particle_in(par);
+        vtx->add_particle_out(par);
```

Comment:
I'm a little confused about the vertices. My (limited) understanding was that a vertex represents an interaction point, where the in/out particles are the incoming and outgoing particles in the interaction (e.g. a decay with one incoming and multiple outgoing particles). Here it looks like we're creating a new vertex for each new position and setting the particle as its own parent?

Reply:
HepMC3 fails if you just have "particle out" (which is what we're getting from Geant4) without an "in". Maybe we also need to have a dummy "particle in"?